src/monitor/group_state_machine.c

```diff
@@ -570,6 +788,611 @@ ProceedGroupStateForPrimaryNode(AutoFailoverNode *primaryNode)
+
+		/* we might not have a selected candidate for failover yet */
+		if (selectedNode == NULL)
+		{
+			/*
+			 * Publish more information about the process in the monitor event
+			 * table. This is a quite complex mechanism here, and it should be
+			 * made as easy as possible to analyze and debug.
+			 */
+			char message[BUFSIZE] = { 0 };
+
+			LogAndNotifyMessage(
+				message, BUFSIZE,
+				"Failover still in progress after all %d candidate nodes "
+				"reported their LSN and we failed to select one of them; "
+				"activeNode is %d (%s:%d) and reported state \"%s\"",
+				candidateList.candidateCount,
+				activeNode->nodeId, activeNode->nodeHost, activeNode->nodePort,
+				ReplicationStateGetName(activeNode->reportedState));
+
+			return false;
+		}
+
+		return PromoteSelectedNode(selectedNode,
+								   primaryNode,
+								   &candidateList);
+	}
+
+	return false;
+}
+
+
+/*
+ * BuildCandidateList builds the list of current standby candidates that have
+ * already reported their LSN, and sets
+ */
+static bool
+BuildCandidateList(List *nodesGroupList, CandidateList *candidateList)
+{
+	ListCell *nodeCell = NULL;
+	List *candidateNodesGroupList = NIL;
+
+	List *secondaryStates = list_make2_int(REPLICATION_STATE_SECONDARY,
+										   REPLICATION_STATE_CATCHINGUP);
+
+	foreach(nodeCell, nodesGroupList)
+	{
+		AutoFailoverNode *node = (AutoFailoverNode *) lfirst(nodeCell);
+
+		if (node == NULL)
+		{
+			/* shouldn't happen */
+			ereport(ERROR, (errmsg("BUG: node is NULL")));
+			continue;
+		}
+
+		/* skip old and new primary nodes (if a selection has been made) */
+		if (StateBelongsToPrimary(node->goalState))
+		{
+			elog(LOG,
+				 "Skipping candidate node %d (%s:%d), "
+				 "which is a primary (old or new)",
+				 node->nodeId, node->nodeHost, node->nodePort);
+			continue;
+		}
+
+		/* skip unhealthy nodes to avoid having to wait for them to report */
+		if (IsUnhealthy(node))
+		{
+			elog(LOG,
+				 "Skipping candidate node %d (%s:%d), which is unhealthy",
+				 node->nodeId, node->nodeHost, node->nodePort);
+
+			continue;
+		}
+
+		/* grab healthy standby nodes which have reached REPORT_LSN */
+		if (IsCurrentState(node, REPLICATION_STATE_REPORT_LSN))
+		{
+			candidateNodesGroupList = lappend(candidateNodesGroupList, node);
+
+			continue;
+		}
+
+		/* if REPORT LSN is assigned and not reached yet, count that */
+		if (node->goalState == REPLICATION_STATE_REPORT_LSN)
+		{
+			++(candidateList->missingNodesCount);
+
+			continue;
+		}
+
+		/*
+		 * Nodes in SECONDARY or CATCHINGUP states are candidates due to report
+		 * their LSN.
+		 */
+		if (IsStateIn(node->reportedState, secondaryStates) &&
+			IsStateIn(node->goalState, secondaryStates))
+		{
+			char message[BUFSIZE] = { 0 };
+
+			++(candidateList->missingNodesCount);
+
+			LogAndNotifyMessage(
+				message, BUFSIZE,
+				"Setting goal state of node %d (%s:%d) to report_lsn "
+				"to find the failover candidate",
```

Comment:
Optional style improvement suggestion:
Assigning `AssignGoalState ` inside a function named `BuildCandidateList ` seems  a little odd. So, it might be little nicer to move `AssignGoalState ` to another function, which we call right before `BuildCandidateList `, and in `BuildCandidateList `  we don't care about  `secondaryStates `.

Reply:
The current coding allows to have a single loop over all the standby nodes, that's what I like about it. Let's chat about that point in a meeting later.